lasagne/trainer.py

```diff
@@ -0,0 +1,687 @@
+    parameters_reset: bool
+        Indicates if the network parameters were successfully reset to the
+        initial state
+    """
+    def __init__(self, epoch, reason, parameters_reset):
+        super(TrainingFailedException, self).__init__(
+                'Training failed at epoch {}: {}'.format(epoch, reason))
+        self.epoch = epoch
+        self.reason = reason
+        self.parameters_reset = parameters_reset
+
+
+class TrainingResults (object):
+    """
+    `TrainingResults` instance provide the results of training a neural
+    network using a `Trainer` instance.
+
+    Attributes
+    ----------
+    train_results: list
+        Per epoch results of the training function
+    validation_results: list
+        Per epoch results of the evaluation function applied to the validation
+        set
+    test_results: list
+        Per epoch results of the evaluation function applied to the test
+        set; note test set is only evaluated *when* the validation score
+        improves, so test_results will contain `None` for epochs in which
+        no improvement was obtained
+    best_val_epoch: int
+        The index of the epoch at which the best validation score was obtained
+    best_validation_results: list
+        The best validation results obtained (equivalent to
+        `self.validation_results[self.best_val_epoch]`)
+    best_test_results: list
+        The more recent test result obtained (equivalent to
+        `self.test_results[self.best_val_epoch]` or
+        `self.test_results[-1]` if no validation set was used)
+    last_epoch: int
+        The index of the last epoch that was executed; indicates when training
+        stopped if early exit is enabled
+    """
+    def __init__(self, train_results, validation_results, best_val_epoch,
+                 best_validation_results, test_results,
+                 best_test_results, last_epoch):
+        self.train_results = train_results
+        self.validation_results = validation_results
+        self.best_val_epoch = best_val_epoch
+        self.best_validation_results = best_validation_results
+        self.test_results = test_results
+        self.best_test_results = best_test_results
+        self.last_epoch = last_epoch
+
+
+def train(train_set, val_set=None, test_set=None, train_batch_func=None,
+          train_log_func=None, train_epoch_results_check_func=None,
+          train_pass_epoch_number=False, eval_batch_func=None,
+          eval_log_func=None, val_improved_func=None, val_interval=None,
+          batchsize=128, num_epochs=100, min_epochs=None,
+          val_improve_patience=1, val_improve_patience_factor=0.0,
+          epoch_log_func=None, pre_epoch_callback=None,
+          post_epoch_callback=None, progress_iter_func=None,
+          verbosity=VERBOSITY_EPOCH, log_stream=sys.stdout,
+          log_final_result=True, get_state_func=None, set_state_func=None,
+          layer_to_restore=None, updates_to_restore=None, shuffle_rng=None):
```

Comment:
Agree with @benanne, this may be a bit heavy on options. My idea was to have a relatively simple `train` function that calls a `train_epoch` function in a loop, and when you need anything more complex, you'd call the `train_epoch` function yourself and extend the loop body with whatever you need (learning rate updates, early stopping, ...). I'd need more time to think this through, though, and I won't have that time soon. Sorry for delaying this for so long, we certainly appreciate your efforts and I would love to continue this discussion in one or two months!